Mathlib/Algebra/Algebra/Subalgebra/Unitization.lean

```diff
@@ -12,99 +12,181 @@ import Mathlib.Algebra.Star.NonUnitalSubalgebra
   where `s` is a non-unital subalgebra of a unital `R`-algebra `A`, this is the natural algebra
-  homomorphism sending `(r, a)` to `r • 1 + a`. This is always surjective.
-* `NonUnitalSubalgebra.unitizationAlgEquiv s : Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A)`:
-  when `R` is a field and `1 ∉ s`, the above homomorphism is injective is upgraded to
-  an `AlgEquiv`.
-* `NonUnitalSubsemiring.unitization : Unitization ℕ S →ₐ[ℕ] Subsemiring.closure (S : Set R)`:
-  the natural `ℕ`-algebra homomorphism between the unitization of a non-unital subsemiring `S` and
-  its `Subsemiring.closure`. This is just `NonUnitalSubalgebra.unitization S` where we replace the
-  codomain using `NonUnitalSubsemiring.closureEquivAdjoinNat`
-* `NonUnitalSubring.unitization : Unitization ℤ S →ₐ[ℤ] Subring.closure (S : Set R)`:
-  the natural `ℤ`-algebra homomorphism between the unitization of a non-unital subring `S` and
-  its `Subring.closure`. This is just `NonUnitalSubalgebra.unitization S` where we replace the
-  codomain using `NonUnitalSubring.closureEquivAdjoinInt`
-
+  homomorphism sending `(r, a)` to `r • 1 + a`. The range of this map is
+  `Algebra.adjoin R (s : Set A)`.
+* `NonUnitalSubalgebra.unitizationAlgEquiv s : Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A)`
+  when `R` is a field and `1 ∉ s`. This is `NonUnitalSubalgebra.unitization` upgraded to an
+  `AlgEquiv` onto its range.
+* `NonUnitalSubsemiring.unitization : Unitization ℕ s →ₐ[ℕ] R`: the natural `ℕ`-algebra homomorphism
+  from the unitization of a non-unital subsemiring `s` into the ring containing it. The range of
+  this map is `subalgebraOfSubsemiring (Subsemiring.closure s)`.
+  This is just `NonUnitalSubalgebra.unitization s` but we provide a separate declaration because
+  there is an instance Lean can't find on its own due to `outParam`.
+* `NonUnitalSubring.unitization : Unitization ℤ s →ₐ[ℤ] R`:
+  the natural `ℤ`-algebra homomorphism from the unitization of a non-unital subring `s` into the
+  ring containing it. The range of this map is `subalgebraOfSubring (Subring.closure s)`.
+  This is just `NonUnitalSubalgebra.unitization s` but we provide a separate declaration because
+  there is an instance Lean can't find on its own due to `outParam`.
+* `NonUnitalStarSubalgebra s : Unitization R s →⋆ₐ[R] A`: a version of
+  `NonUnitalSubalgebra.unitization` for star algebras.
+* `NonUnitalStarSubalgebra.unitizationStarAlgEquiv s :`
+  `Unitization R s ≃⋆ₐ[R] StarSubalgebra.adjoin R (s : Set A)`:
+  a version of `NonUnitalSubalgebra.unitizationAlgEquiv` for star algebras.
 -/
 
-section Generic
+/-! ## Subalgebras -/
+
+section Subalgebra
+
+variable {R A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
+
+/-- Turn a `Subalgebra` into a `NonUnitalSubalgebra` by forgetting that it contains `1`. -/
+def Subalgebra.toNonUnitalSubalgebra (S : Subalgebra R A) : NonUnitalSubalgebra R A :=
+  { S with
+    smul_mem' := fun r _x hx => S.smul_mem hx r }
```

Comment:
Is this really the earliest place that this can be defined?

Reply:
Yes, because neither the unital nor the non-unital subobject files import each other.